### DISCUSSION TITLE
How to use Monaco Editor in a browser extension?

### DISCUSSION THREAD

**Discussion by erosman** [microsoft/monaco-editor#3908] [Q&A] [answered]
Is it possible to `import` Monaco-editor as ESM directly in a browser extension environment, **without the use of 3rd party packing scripts**?

**Comment by tamayika** [comment, marked as answer]
I think it's not possible because current browser esm cannot import css.
If [import attributes](https://github.com/tc39/proposal-import-attributes) landed in browsers and monaco-editor load css with import attributes, you will be able to use monaco-editor without bundling tools.

**Comment by erosman** [comment]
I think, I should be able to work around CSS issue.

I was more concerned about having to include npm (or other) dependencies.

I am currently using CodeMirror 5 in my Firefox extension which can be used directly, but CodeMIrror 6 needs to be compiled to use.

I use VS Code for my own programming and was hoping to use it in my Firefox extension as well.

I would only need JavaScript & CSS processing for the editor.

**Reply by renish-c** [reply]
Hey @erosman,
Were you able to work around css import issue. 

If possible can you drop a guide on how can we use monaco-editor in chrome extension. We used it but it is throwing CDN issues, we can add cdn path to manifest.json to allow it in CSP. But I was wondering if there is any way we could use everything locally to avoid cdn issues thrown by chrome extension content scripts.

**Reply by DavidSeptimus** [reply]
A few notes from my own experience building [Tachi Code](https://tachicode.com):
- Downloading monaco via CDN is going to get your extension flagged by Chrome. You're going to need package the editor with your extension.
- I use @monaco-editor/react and had to patch @monaco-editor/loader to remove the code that falls back to downloading monaco from a CDN to get approved despite my extension not actually calling it.
- The typescript worker as packaged is too large for Firefox (I think it's something like 12mb and firefox has a file size limit of 5mb). I was able to shrink it down by copying the original typescript source into my project and using vite/rollup to chunk and minify it down to an acceptable size

**Reply by erosman** [reply]
@renish-c Browsers (Chrome, Firefox, etc) do not allow remote JavaScript, therefore using a remote CDN is not possible in a browser extension.

You can grab all the necessary files and pack them with the extension but each file has to match the checksum of the official original which is complicated and bulky.

I used the examples in [samples](https://github.com/microsoft/monaco-editor/tree/main/samples) to compile a bundle and it is working fine in the browser environment.
The final outcome is still 11.9 MB unzipped but once zipped into an extension, it becomes 2.9 MB.

**Reply by erosman** [reply]
@DavidSeptimus Firefox has a 4 MB size limit for each background script file. Monaco-editor wont be running as a background script, so there shouldn't be an issue.

**Reply by DavidSeptimus** [reply]
@erosman I hit this issue myself when I first tried releasing my own add-on. It's a JS file size limitation due to performance constraints of their internal JS linter for add-ons. It applies to any JS file bundled in your extension, not just background scripts.

It's been an ongoing discussion for years and it looks like [they only recently raised the cap from 4MB to 5MB](https://github.com/mozilla/addons-linter/issues/4942). As far as I'm aware, the compressed size doesn't matter in this instance since they need to decompress your extension to lint the JS anyhow.

**Reply by erosman** [reply]
@DavidSeptimus You are right. I tested it now on [validator](https://addons.mozilla.org/en-US/developers/addon/validate) and although Firefox has no problem handling the files, [addons-linter](https://github.com/mozilla/addons-linter) complains as the linter has issues parsing the file. I am going to talk with the dev team about it.

**Reply by DavidSeptimus** [reply]
Back to the topic of the actual file size of the TS worker:
- ESM:  **11.2mb**
- Minified AMD: **5.7mb**

So it definitely needs to be chunked for firefox at the moment and I know my experience is limited to my knowledge of chunking with Vite + Rollup. I know I couldn't figure out how to force it to chunk a JS file from an npm dependency. Could be there's a way to make that work that I'm not aware of.

Alternatively, maybe there's a way to do it with Webpack, which, to my knowledge, offers greater control over code splitting. That said, I've never really done much with Webpack, so I couldn't speak too much to what's possible there without more investigation.

If someone more experienced with either bundler has any ideas beyond my original suggestion (which I use in production), please share (I'd love to be able to drop the TS worker from my codebase).

**Reply by erosman** [reply]
See also:

- [Bundling Libraries filesize and addons-linter](https://github.com/mozilla/addons-linter/issues/5821)
- [[Feature Request] Split typescript service-worker into separate modules](https://github.com/microsoft/monaco-editor/issues/5061)
- [Use LSP for TypeScript language features](https://github.com/microsoft/monaco-editor/issues/5065)
